crates/sui-framework/packages/dwallet-system/sources/ethereum_state.move

```diff
@@ -0,0 +1,112 @@
+/*
+ * ethereum_state module provides functionality for managing and verifying the state of Ethereum within
+ * the dwallet_system. It includes structures to represent Ethereum state, functions to initialize and update state,
+ * and native functions to verify state updates and create initial state data.
+*/
+module dwallet_system::ethereum_state {
+    use dwallet::object::{Self, UID, ID};
+    use dwallet::transfer;
+    use dwallet::tx_context::TxContext;
+
+    /// Ethereum state object.
+    struct EthereumState has key {
+        id: UID,
+        /// serialised ConsensusStateManager
+        data: vector<u8>,
+        time_slot: u64,
+    }
+
+    /// Latest Ethereum state object.
+    /// Holds the ID of the latest Ethereum state object that is verified by the dWallet network.
+    struct LatestEthereumState has key, store {
+        id: UID,
+        eth_state_id: ID,
+        last_slot: u64,
+        eth_smart_contract_addr: vector<u8>,
+        eth_smart_contract_slot: u64,
+    }
+
+    /// Initializes the first Ethereum state with the given checkpoint.
+    /// Creates an EthereumState object, shares a LatestEthereumState object pointing to it,
+    /// and freezes the EthereumState object.
+    public fun init_state(
+        checkpoint: vector<u8>,
+        network: vector<u8>,
+        eth_smart_contract_addr: vector<u8>,
+        eth_smart_contract_slot: u64,
+        ctx: &mut TxContext
+    ) {
+        let state_data = create_initial_eth_state_data(checkpoint, network);
+        let state = EthereumState {
+            id: object::new(ctx),
+            data: state_data,
+            time_slot: 0u64,
+        };
+
+        transfer::share_object(LatestEthereumState {
+            id: object::new(ctx),
+            eth_state_id: object::id(&state),
+            last_slot: state.time_slot,
+            eth_smart_contract_addr,
+            eth_smart_contract_slot,
+        });
+        transfer::freeze_object(state);
+    }
+
+    /// Updates the latest Ethereum state reference if the provided EthereumState object has a newer time slot.
+    public fun update_latest_eth_state(
+        self: &mut LatestEthereumState,
+        eth_state: &EthereumState,
+    ) {
+        if (eth_state.time_slot > self.last_slot) {
```

Comment:
Here I'm missing a network check: so that updates for Ethereum `testnet` could not be set to the state of the `mainnet` object.